mllib/src/main/scala/org/apache/spark/ml/regression/AFTRegression.scala

```diff
@@ -0,0 +1,308 @@
+    if (fitting) {
+      SchemaUtils.checkColumnType(schema, $(censorCol), DoubleType)
+      SchemaUtils.checkColumnType(schema, $(labelCol), DoubleType)
+    } else {
+      SchemaUtils.checkColumnType(schema, $(quantileCol), new VectorUDT)
+    }
+    SchemaUtils.appendColumn(schema, $(predictionCol), DoubleType)
+  }
+}
+
+@Experimental
+class AFTRegression(override val uid: String)
+  extends Estimator[AFTRegressionModel] with AFTRegressionParams with Logging {
+
+  def this() = this(Identifiable.randomUID("aftReg"))
+
+  /** @group setParam */
+  def setFeaturesCol(value: String): this.type = set(featuresCol, value)
+
+  /** @group setParam */
+  def setLabelCol(value: String): this.type = set(labelCol, value)
+
+  /** @group setParam */
+  def setCensorCol(value: String): this.type = set(censorCol, value)
+  setDefault(censorCol -> "censored")
+
+  /** @group setParam */
+  def setPredictionCol(value: String): this.type = set(predictionCol, value)
+
+  /**
+   * Set if we should fit the intercept
+   * Default is true.
+   * @group setParam
+   */
+  def setFitIntercept(value: Boolean): this.type = set(fitIntercept, value)
+  setDefault(fitIntercept -> true)
+
+  /**
+   * Set the maximum number of iterations.
+   * Default is 100.
+   * @group setParam
+   */
+  def setMaxIter(value: Int): this.type = set(maxIter, value)
+  setDefault(maxIter -> 100)
+
+  /**
+   * Set the convergence tolerance of iterations.
+   * Smaller value will lead to higher accuracy with the cost of more iterations.
+   * Default is 1E-6.
+   * @group setParam
+   */
+  def setTol(value: Double): this.type = set(tol, value)
+  setDefault(tol -> 1E-6)
+
+  /**
+   * Extracts (feature, label, censored) from input dataset.
+   */
+  protected[ml] def extractCensoredLabeledPoints(
+      dataset: DataFrame): RDD[(Vector, Double, Double)] = {
+    dataset.select($(featuresCol), $(labelCol), $(censorCol))
+      .map { case Row(feature: Vector, label: Double, censored: Double) =>
+      (feature, label, censored)
+    }
+  }
+
+  override def fit(dataset: DataFrame): AFTRegressionModel = {
+    validateAndTransformSchema(dataset.schema, fitting = true)
+    val instances = extractCensoredLabeledPoints(dataset)
+    val handlePersistence = dataset.rdd.getStorageLevel == StorageLevel.NONE
+    if (handlePersistence) instances.persist(StorageLevel.MEMORY_AND_DISK)
+
+    val costFun = new AFTCostFun(instances, $(fitIntercept))
+    val optimizer = new BreezeLBFGS[BDV[Double]]($(maxIter), 10, $(tol))
+
+    val numFeatures = dataset.select($(featuresCol)).take(1)(0).getAs[Vector](0).size
+    /*
+       The weights vector has three parts:
+       the first element: Double, log(sigma), the log of scale parameter
+       the second element: Double, intercept of the beta parameter
+       the third to the end elements: Doubles, weights vector of the beta parameter
+     */
+    val initialWeights = new DenseVector(Array.fill(numFeatures + 2)(1.0))
```

Comment:
here intercept (2nd element) is initialized to 1 regardless of `fitIntercept`. If `fitIntercept` is false, the 2nd element will stay as 1 after the LBFGS optimization. This value will be carry to the `AFTRegressionModel`. Then `predict` and `transform` will be incorrect. 

How about we include intercept in `initialWeights` *only* when `fitIntercept` is `true`?

Btw, shall we set the initial values to be zero since we are working in the log scale.